### ISSUE TITLE
Feature request: ML-backed prompt router using profile feature vectors

### ISSUE THREAD

**Issue by zavera** [iryna-kondr/scikit-llm#141]
## Summary

I am working on a structured computation pipeline where LLM outputs vary significantly based on input profile characteristics. The core problem is that no single prompt performs best across all input profiles. Different prompt versions win on different profile types, and the winning prompt is predictable from a small set of boolean features extracted from the input.

I would like to propose a `ProfileAwareRouter` component that bridges classical ML routing with LLM prompt selection. This fits naturally into scikit-LLM's mission of combining scikit-learn patterns with LLM workflows.

## The Pattern

Given a structured input payload, extract a boolean feature vector representing profile characteristics (for example, which forms or schema elements are present). Use KNN or a Random Forest trained on a labeled evaluation matrix to route the input to the prompt version most likely to produce a correct output.

```python
from skllm.routing import ProfileAwareRouter

router = ProfileAwareRouter(
    prompts={"v1": prompt_v1, "v2": prompt_v2, "v3": prompt_v3},
    feature_extractor=extract_boolean_flags,
    routing_model="knn",          # or "random_forest"
    ensemble_fields=["net_biz"]   # fields where no single prompt wins, use weighted ensemble
)

router.fit(X_profile_features, y_correct_prompt_labels)
result = router.predict(new_payload)
```

## Key Components

**Profile feature extraction**
Extract a low-dimensional boolean feature vector from any structured payload. Keeps dimensionality low enough for KNN distance to be meaningful on small labeled datasets (200-500 records is realistic in domain-specific pipelines).

**Elbow-based cluster discovery**
Automatically select k using inertia elbow method with silhouette score validation. Reveals natural profile groupings in the labeled dataset that correspond to meaningful input types.

**KNN routing**
Match new inputs to their nearest neighbors in profile space. Return the winning prompt for that neighborhood along with a confidence score based on neighbor agreement.

**Random Forest ensemble for ambiguous fields**
For fields where no single prompt dominates, use a Random Forest where all prompt outputs plus profile features are input features, and the forest learns to reconcile them into a final prediction. Includes out-of-bag error estimation for honest accuracy on small datasets.

**Confidence gating**
When match confidence falls below a threshold, fall back to weighted ensemble rather than hard routing. Flag low confidence records for human review.

## Why This Belongs in scikit-LLM

This pattern generalizes beyond any single domain. Anywhere that structured payloads feed LLM computation and prompt performance varies by input characteristics, this router applies. Tax computation, medical coding, insurance underwriting, financial aid calculation are all examples.

The implementation would follow scikit-learn conventions (fit/predict API, compatible with Pipeline) making it immediately familiar to the existing scikit-LLM user base.

## Contribution Interest

I am actively building this pattern and would be interested in contributing a reference implementation if the maintainers think it fits the project direction. Happy to discuss architecture before writing code.

**Comment by modelpath-dev** [comment]
I will take this task. Please assign it to me.

The issue suggests adding a `ProfileAwareRouter` to select prompts based on input profile features. I would start by examining how the existing routing components are structured in the codebase. Then, I would implement a new class that integrates with scikit-LLM, using either KNN or Random Forest for routing. The first step would be to create a feature extractor for boolean vectors and ensure it works with the existing data structures. I would then test the routing logic with a small dataset to verify the correct prompt selection.

**Comment by modelpath-dev** [comment]
Could you please assign this issue to me? I will start by reviewing the current routing components and then work on implementing the `ProfileAwareRouter` class. My initial focus will be on developing the feature extractor for boolean vectors and testing the routing logic with a small dataset. Let me know if there are any specific requirements or considerations to keep in mind.

**Comment by OKUA1** [comment]
Hi @zavera, 

Thanks for the writeup. The motivation makes sense to me, but the design is still pretty vague and I have a bunch of questions before we can talk about whether this fits. 

For example:

- Where do the labels come from? y_correct_prompt_labels implies you already ran every prompt on every record and scored them against ground truth. Is that meant to happen outside skllm, or is an evaluation step part of the proposal? If outside, what does the input look like?

- Should the fitting even be in skllm? The routing model is just a sklearn classifier on features the user extracts themselves. What does skllm add beyond a KNeighborsClassifier and a dict of prompts? Maybe it's enough to take an already fitted estimator and only handle the dispatch.

- What does predict return, the prompt key or the LLM output? Also, fit takes extracted features but predict takes a raw payload, which seems inconsistent.

- Are the prompts free form or do they need to follow a template? For the RF reconciliation step the outputs have to be comparable, so I'd assume a shared output schema is required.

- The RF ensemble and the low confidence fallback both seem to need all prompts run at inference. Is that intended? It kind of goes against the point of routing to a single prompt.

- ensemble_fields suggests winners differ per output field, but the router picks one prompt per payload. How do those two fit together?

**Comment by modelpath-dev** [comment]
These are great questions. For the labels, the idea is that they would be generated externally by running each prompt on the dataset and scoring against ground truth. The fitting could indeed be done outside skllm, with skllm focusing on dispatch. As for `predict`, it should return the prompt key, and the input should be consistent between `fit` and `predict`. Prompts would need a shared output schema for comparability. The RF ensemble and low confidence fallback would indeed require all prompts to be run at inference time.
